File: node/perftestmongo.js
const MongoClient = require('mongodb').MongoClient;
const url = "mongodb://db1.cams";
 
function time() {
  return new Date().getTime();
}

function cpuTime() {
  const cpu = process.cpuUsage();
  return (cpu.user + cpu.system) / 1000;
}

function getAverage(data) {
  data = data.slice(10);
  data.sort();
  let result = 0;
  for(let i = 0; i < data.length - 2; i++) result += data[i];
  return result / (data.length - 2);
} 
 
(async () => {
    try {
        // make sure that any items are correctly URL encoded in the connection string
        const client = await MongoClient.connect(url);
        const db = client.db("test");
        const collection = db.collection("tsdata");
        const totalRows = 100000;
        let first = true; 
        let inserts = [];
        let updates = [];
        let selects = [];
        let insertsCpu = [];
        let updatesCpu = [];
        let selectsCpu = [];
        
        for(let i = 0; i < 21; i++) {
          // Clear
          await collection.drop();
          global.gc();
          
          // Insert
          let start = time();
          let startCpu = cpuTime();
            
          let data = [];
          for(let i = 0; i < totalRows; i++) {
            data.push({_id: i, created: new Date(), value: i / 10.0});
          }
          await collection.insertMany(data);
          global.gc();
          console.log(`Time to insert: ${time() - start} ${cpuTime() - startCpu}`);
          inserts.push(time() - start);
          insertsCpu.push(cpuTime() - startCpu);

          
          // Update
          start = time();
          startCpu = cpuTime();
          let ops = [];
          for(let i = 0; i < totalRows; i++) {
            ops.push({updateOne: { filter: { _id: i}, update: {"$set": {value: i / 5.0}}}});
          }
          await collection.bulkWrite(ops);
          global.gc();
          
          console.log(`Time to update: ${time() - start} ${cpuTime() - startCpu}`);
          updates.push(time() - start);
          updatesCpu.push(cpuTime() - startCpu);
          
          // Select          
          start = time();
          startCpu = cpuTime();
          const result = await collection.find().toArray();
          const rows = [];
          result.forEach(item => rows.push({_id: item._id, created: item.created, value: item.value}));
          global.gc();

          console.log(`Selected ${rows.length} in ${time() - start} ${cpuTime() - startCpu}`);
          selects.push(time() - start);
          selectsCpu.push(cpuTime() - startCpu);
        }
        
        console.log(`Average insert: ${getAverage(inserts)} ${getAverage(insertsCpu)}`);
        console.log(`Average update: ${getAverage(updates)} ${getAverage(updatesCpu)}`);
        console.log(`Average select: ${getAverage(selects)} ${getAverage(selectsCpu)}`);
    } catch (err) {
        console.log(err);
    }
})()

File: java/src/main/java/PerfTestMongo.java
import com.mongodb.MongoClient;
import com.mongodb.client.model.*;
import org.bson.codecs.pojo.PojoCodecProvider;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;

import static com.mongodb.MongoClientSettings.getDefaultCodecRegistry;
import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;


public class PerfTestMongo {

    public static void main(String[] args) throws IOException {
        var totalRows = 100000;
        var inserts = new ArrayList<Long>();
        var batchUpdates = new ArrayList<Long>();
        var selects = new ArrayList<Long>();
        var insertsCpu = new ArrayList<Long>();
        var batchUpdatesCpu = new ArrayList<Long>();
        var selectsCpu = new ArrayList<Long>();

        // Register codec so that we can communicate using POJOs
        var pojoCodecProvider = PojoCodecProvider.builder().automatic(true).register(Item.class).build();
        var pojoCodecRegistry = fromRegistries(getDefaultCodecRegistry(), fromProviders(pojoCodecProvider));
        var mongoClient = new MongoClient("db1.cams");
        var db = mongoClient.getDatabase("test").withCodecRegistry(pojoCodecRegistry);

        for (var it = 0; it < 21; it++) {
            // Remove all data
            var collection = db.getCollection("tsdata").withDocumentClass(Item.class);
            collection.drop();
            System.gc();

            // Insert rows
            var sw = StopWatch.startNew();
            var data = new ArrayList<Item>(totalRows);
            for (var i = 0; i < totalRows; i++) {
                var item = new Item(i, new Date(1000000), ((double) i / 10));
                data.add(item);
            }
            collection.insertMany(data, new InsertManyOptions().ordered(false));
            System.gc();
            System.out.println("Time to insert: " + sw.elapsed() + " " + sw.cpuElapsed());
            inserts.add(sw.elapsed());
            insertsCpu.add(sw.cpuElapsed());

            // Update rows using JDBC batch update
            sw = StopWatch.startNew();
            var writes = new ArrayList<WriteModel<Item>>(totalRows);
            for (var i = 0; i < totalRows; i++) {
                writes.add(new UpdateOneModel<>(Filters.eq("_id", i), Updates.set("value", (double) i / 5)));
            }
            collection.bulkWrite(writes, new BulkWriteOptions().ordered(false));
            System.gc();
            System.out.println("Time to batch update: " + sw.elapsed() + " " + sw.cpuElapsed());
            batchUpdates.add(sw.elapsed());
            batchUpdatesCpu.add(sw.cpuElapsed());

            // Select rows
            sw = StopWatch.startNew();
            var cursor = collection.find(Item.class);
            var rows = new ArrayList<Item>(totalRows);
            for (var item : cursor) {
                rows.add(item);
            }
            System.gc();
            System.out.println("Selected " + rows.size() + " in " + sw.elapsed() + " " + sw.cpuElapsed());
            selects.add(sw.elapsed());
            selectsCpu.add(sw.cpuElapsed());
        }

        System.out.println("Average insert: " + PerfTest.getAverage(inserts) + " " + PerfTest.getAverage(insertsCpu));
        System.out.println("Average batch update: " + PerfTest.getAverage(batchUpdates) + " " + PerfTest.getAverage(batchUpdatesCpu));
        System.out.println("Average select: " + PerfTest.getAverage(selects) + " " + PerfTest.getAverage(selectsCpu));

        System.in.read();

    }
}
